src/v/cluster/jumbo_log_stm.cc

```diff
@@ -0,0 +1,221 @@
+    auto write_intent_id_it = _write_intent_id_by_object_id.find(object.id);
+    if (write_intent_id_it != _write_intent_id_by_object_id.end()) {
+        // Ensure that size_bytes also matches and the client is not reusing
+        // object ids.
+        if (auto write_intent_it = _write_intents.find(
+              write_intent_id_it->second);
+            write_intent_it != _write_intents.end()
+            && write_intent_it->second.object.size_bytes != object.size_bytes) {
+            throw std::system_error(
+              raft::make_error_code(raft::errc::invalid_input_records));
+        }
+        co_return write_intent_id_it->second;
+    }
+
+    storage::record_batch_builder batch_builder(
+      model::record_batch_type::jumbo_log_cmd, model::offset{});
+
+    auto intent_id = _next_write_intent_id;
+
+    auto record = create_write_intent_cmd::value{
+      .segment = jumbo_log::write_intent_segment(intent_id, object)};
+
+    batch_builder.add_raw_kv(
+      serde::to_iobuf(create_write_intent_cmd::key), serde::to_iobuf(record));
+
+    // No timeouts here! We can't allow concurrency as it will break write
+    // intent id allocation invariants.
+    auto r = co_await _raft->replicate(
+      _insync_term,
+      model::make_memory_record_batch_reader(std::move(batch_builder).build()),
+      raft::replicate_options(raft::consistency_level::quorum_ack));
+    if (!r) {
+        throw std::system_error(raft::make_error_code(raft::errc::timeout));
+    }
+
+    // TODO(nv): Sure ok to timeout here?
+    if (!co_await wait_no_throw(
+          model::offset(r.value().last_offset()),
+          model::timeout_clock::now() + timeout)) {
+        throw std::system_error(raft::make_error_code(raft::errc::timeout));
+    }
+
+    co_return intent_id;
+}
+
+ss::future<bool> jumbo_log_stm::sync(model::timeout_clock::duration timeout) {
+    auto is_synced = co_await persisted_stm::sync(timeout);
+
+    co_return is_synced;
+}
+
+ss::future<> jumbo_log_stm::apply(const model::record_batch& batch) {
+    if (batch.header().type != model::record_batch_type::jumbo_log_cmd) {
+        co_return;
+    }
+
+    batch.for_each_record([this](model::record&& r) {
+        auto key = serde::from_iobuf<cmd_key>(r.release_key());
+
+        switch (key) {
+        case create_write_intent_cmd::key: {
+            auto intent = serde::from_iobuf<create_write_intent_cmd::value>(
+              r.release_value());
+
+            apply_create_write_intent(std::move(intent));
+            break;
+        }
+        default:
+            throw std::runtime_error(fmt_with_ctx(
+              fmt::format,
+              "Unknown jumbo_log_stm command {}",
+              static_cast<int>(key)));
+        }
+
+        return ss::stop_iteration::no;
+    });
+
+    co_return;
+}
+
+ss::future<> jumbo_log_stm::write_snapshot() {
+    if (_is_writing_snapshot) {
+        return ss::now();
+    }
```

Comment:
just curious if this is a pattern you're copying from other STMs or something specific to JL?

Reply:
copying from the id allocator stm mostly